### ISSUE TITLE
CPU offloading

### ISSUE THREAD

**Issue by LiliumSancta** [mlc-ai/mlc-llm#5]
Incredible project, i managed to run the model with good speed on my hardware (AMD) thanks. 
I have a question do you have any plans to offload the weights and be able to run bigger models like 13B or 30B with less vram?

**Comment by junrushao** [comment]
Hey thanks for your interest! Our backend (TVM Unity) supports AMD CPUs out of box, so it wouldn’t be too challenging (likely tens of lines) to introduce support for them. Not too sure about the latency aspect though

**Comment by Dampfinchen** [comment]
Yeah, the limitations current LLM inference programs like Oobabooga WebUI and KoboldAi have is that CPU offloading is very slow for these. Maybe the MLC team could build a very fast CPU offloader that allocates RAM on the fly as soon as VRAM is overflowing, to prevent out of memory errors at high context sizes and with big models.

**Comment by AlphaAtlas** [comment]
The dream (for my hardware) is being able to split up the model between separate vulkan devices... maybe splitting up layers like llama.cpp does? This would allow for hybrid IGP+GPU inference. 

Splitting the model between multiple backends is probably outside the domain of tvm though, right?

**Comment by authwork** [comment]
@yzh119 @AlphaAtlas 
Agreed, and is there a way to support multi-threading model lib generation for mlc-llm?
```
python3 build.py --model /home/xyao/data/vicuna-7b-v1.3
./build/mlc_chat_cli --local-id vicuna-7b-v1.3-q0f16 --device-name cpu
```
it seems can only run the model inference via one cpu core, and it is a little bit hard to compare the performance between mlc-llm and llama.cpp.

**Comment by sirus20x6** [comment]
as someone that currently does cpu inference I would love this feature.

**Comment by MikeLP** [comment]
Any updates?

**Comment by mengshyu** [comment]
The initial CPU support has already been merged(https://github.com/mlc-ai/mlc-llm/pull/3116 ). Next, we will be working on expanding model coverage and optimizing performance.
